base/strings/substring.jl

```diff
@@ -162,10 +145,16 @@ julia> join(reverse(collect(graphemes("ax̂e")))) # reverses graphemes
 "ex̂a"
 ```
 """
-reverse(s::AbstractString) = RevString(s)
-reverse(s::RevString) = s.string
-
-## reverse an index i so that reverse(s)[i] == s[reverseind(s,i)]
+function reverse(s::S)::S where {S<:AbstractString}
+    # TODO: generic API for sprint to a particular encoding
```

Comment:
I'm not sure whether it's a good idea to have this generic `reverse` method or not. It does ensure that the reversed string type is the same as the input string type, but it doesn't do so in a particularly efficient way. Calling `sprint` with a string encoding based on the input string type could make this both generic and efficient, but we don't currently have a way to do that in Base. Thoughts, @nalimilan, @stevengj?

Reply:
I don't understand how this implementation ensures that the reversed string type is the same as the input string type.   Doesn't `sprint()` always return `String`?

Reply:
I'd just implement this for `String` and `SubString`, and leave it for custom implementations to implement it. We can always add a generic method after 1.0 if we want.

Reply:
It has a return type annotation which implicitly calls convert for you. Perhaps too subtle?

Reply:
Oh, I missed that; still not used to looking for return-type annotations.  +1 for omitting this fallback entirely.